File: backend/app/utils/thumbnail_sort.py
"""
Utility functions for sorting thumbnail URLs by ordinal.
Thumbnail format: [postid]-t-[ordinal]-[uuid].ext
Example: 122097057-t-000-1b73c5f0.png
"""

import re
from typing import List


def extract_ordinal(url: str) -> int:
    """
    Extract ordinal from thumbnail filename.

    Args:
        url: Thumbnail URL or filename

    Returns:
        Ordinal number, or float('inf') if not found (sorts to end)
    """
    if not url:
        return float('inf')

    # Extract filename from URL
    filename = url.split('/')[-1]

    # Match pattern: anything-t-[digits]-anything
    match = re.search(r'-t-(\d+)-', filename)

    if match:
        return int(match.group(1))

    # If pattern doesn't match, return infinity (sorts to end)
    return float('inf')


def sort_thumbnails(thumbnail_urls: List[str]) -> List[str]:
    """
    Sort thumbnail URLs by ordinal.
    Maintains original order if ordinals can't be extracted.

    Args:
        thumbnail_urls: Array of thumbnail URLs

    Returns:
        Sorted array
    """
    if not thumbnail_urls:
        return thumbnail_urls

    # Create list of (url, ordinal, original_index) tuples
    with_ordinals = [
        (url, extract_ordinal(url), idx)
        for idx, url in enumerate(thumbnail_urls)
    ]

    # Sort by ordinal, then by original index (stable sort)
    with_ordinals.sort(key=lambda x: (x[1], x[2]))

    # Return sorted URLs
    return [item[0] for item in with_ordinals]

File: frontend/src/utils/thumbnailSort.js
/**
 * Extract ordinal from thumbnail filename
 * Format: [postid]-t-[ordinal]-[uuid].ext
 * Example: 122097057-t-000-1b73c5f0.png -> 0
 *
 * @param {string} url - Thumbnail URL or filename
 * @returns {number} - Ordinal number, or Infinity if not found (sorts to end)
 */
export function extractOrdinal(url) {
  if (!url) return Infinity;

  // Extract filename from URL
  const filename = url.split("/").pop();

  // Match pattern: anything-t-[digits]-anything
  const match = filename.match(/-t-(\d+)-/);

  if (match && match[1]) {
    return parseInt(match[1], 10);
  }

  // If pattern doesn't match, return Infinity (sorts to end)
  return Infinity;
}

/**
 * Sort thumbnail URLs by ordinal
 * Maintains original order if ordinals can't be extracted
 *
 * @param {Array<string>} thumbnailUrls - Array of thumbnail URLs
 * @returns {Array<string>} - Sorted array
 */
export function sortThumbnails(thumbnailUrls) {
  if (!thumbnailUrls || thumbnailUrls.length === 0) {
    return thumbnailUrls;
  }

  // Create array of [url, ordinal, originalIndex] tuples
  const withOrdinals = thumbnailUrls.map((url, idx) => ({
    url,
    ordinal: extractOrdinal(url),
    originalIndex: idx,
  }));

  // Sort by ordinal, then by original index (stable sort)
  withOrdinals.sort((a, b) => {
    if (a.ordinal !== b.ordinal) {
      return a.ordinal - b.ordinal;
    }
    return a.originalIndex - b.originalIndex;
  });

  // Return sorted URLs
  return withOrdinals.map((item) => item.url);
}
